### PR TITLE
tedge-agent creates /var/tedge/cache directory on start-up

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
This directory was introduced by c8y-firmware-plugin, however, now cloud mappers need to download files from remote. The cache directory must be used also by cloud mappers to avoid multiple file downloads for the same file.

The directory usage will be covered by robot tests during #2292 work.

In addition, to solve this comment https://github.com/thin-edge/thin-edge.io/pull/2293#discussion_r1338429388, I introduced `DataDir`(crates/core/tedge_api/src/path.rs) struct, which is a wrapper of `Utf8PathBuf` with refactoring codes.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2292 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/src/lib.rs [hunk 1]
```diff
@@ -32,6 +32,7 @@ pub use software::*;
 
 pub const DEFAULT_DATA_PATH: &str = "/var/tedge";
 pub const DEFAULT_FILE_TRANSFER_DIR_NAME: &str = "file-transfer";
+pub const DEFAULT_FILE_CACHE_DIR_NAME: &str = "cache";
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/lib.rs:35]: I would really avoid to add yet another `pub const`.  

These constants are too low level. Look at the current usages for `DEFAULT_FILE_TRANSFER_DIR_NAME`. All of these are starting with a `data_dir.join()`. i.e. doing the same as `file_transfer_dir_as_pathbuf()` but with some possibly buggy `data_dir`.

What we need are methods providing paths ready to be used.

**Review by didier-wenzek — COMMENTED** [review]
I think it's better to go a step further by deprecating the 3 `pub const` related to file transfer (namely `DEFAULT_DATA_PATH`, `DEFAULT_FILE_TRANSFER_DIR_NAME` and `DEFAULT_FILE_CACHE_DIR_NAME`).

File: crates/core/tedge_api/src/path.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use std::path::Path;
+
+/// Representation of ThinEdge data directory.
+/// Default is /var/tedge.
+/// All directories under the root data directory must be listed here.
+#[derive(Debug, Clone)]
+pub struct DataDir(pub Utf8PathBuf);
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/path.rs:9]: Do we need `self.0` to be `pub`?

**Review by didier-wenzek — APPROVED** [review]
Approved. Nice improvement!

File: crates/extensions/c8y_config_manager/src/config.rs [hunk 3]
```diff
@@ -49,7 +49,7 @@ impl ConfigManagerConfig {
         let plugin_config_dir = config_dir.join(DEFAULT_OPERATION_DIR_NAME);
         let plugin_config_path = plugin_config_dir.join(DEFAULT_PLUGIN_CONFIG_FILE_NAME);
 
-        let file_transfer_dir = data_dir.join(DEFAULT_FILE_TRANSFER_DIR_NAME);
+        let file_transfer_dir = data_dir.file_transfer_dir().as_std_path().to_path_buf();
```
Comment (albinsuresh) [inline crates/extensions/c8y_config_manager/src/config.rs:52]:
```suggestion
        let file_transfer_dir = data_dir.file_transfer_dir().as_std_path().to_path_buf();
```
```suggestion
        let file_transfer_dir = data_dir.file_transfer_dir().into();
```

File: crates/extensions/c8y_firmware_manager/src/actor.rs [hunk 4]
```diff
@@ -253,13 +253,13 @@ impl FirmwareManagerActor {
                 .is_some()
             {
                 if let Ok(token) = self.message_box.jwt_retriever.await_response(()).await? {
-                    DownloadRequest::new(firmware_url, &cache_file_path)
+                    DownloadRequest::new(firmware_url, cache_file_path.as_std_path())
```
Comment (albinsuresh) [inline crates/extensions/c8y_firmware_manager/src/actor.rs:256]:
```suggestion
                    DownloadRequest::new(firmware_url, cache_file_path.as_ref())
```
Not a request for change but more of an FYI. But, I also know that you prefer explicit conversion functions over the generic ones like `as_ref()` or `into()` 😄

**Review by albinsuresh — COMMENTED** [review]
The core change looks fine. But, as a result of this introduction of the `DataDir` type, with helper methods that returns `Utf8PathBuf` values for things like `file_transfer_repo_path`, you ended changing a lot of code which was accepting the standard `&Path` or `PathBuf` fields to accept `impl AsRef<Path>` or directly `Utf8PathBuf`. You probably didn't need to change all of that if `DataDir` had all the necessary conversion impls like `impl From<DataDir> for PathBuf`,` impl From<DataDir> for Utf8PathBuf` etc. Seemless conversion of a DataDir into all the standard path types should have been possible with these `From` impls and the `AsRef<Path>` impl that you already had.

That being said, using `Utf8PathBuf` and `Utf8Path` across the board is not such a bad idea either. I'm just saying that you probably didn't **have to** make all these changes just because `DataDir` returns `Utf8PathBuf`s.

**Comment by rina23q** [comment]
> That being said, using Utf8PathBuf and Utf8Path across the board is not such a bad idea either.

Since `tedge_config` API returns `Utf8PathBuf`, using `PathBuf` is a kind of downgrade in most of cases (only config_path from command line option is different, I'm not sure if clap can accept `Utf8Pathbuf`.)

So, I even wish I could change all usage of `PathBuf` to `Utf8PathBuf` in several crates.

File: crates/core/tedge_api/src/path.rs (hunk 2 above)
Comment (rina23q) [reply crates/core/tedge_api/src/path.rs:9]: I'll add it as @albinsuresh suggested.

File: crates/extensions/c8y_firmware_manager/src/config.rs [hunk 5]
```diff
@@ -19,11 +20,11 @@ const FIRMWARE_UPDATE_RESPONSE_TOPICS: &str = "tedge/+/commands/res/firmware_upd
...
-    pub tmp_dir: PathBuf,
-    pub data_dir: PathBuf,
-    pub cache_dir: PathBuf,
-    pub file_transfer_dir: PathBuf,
-    pub firmware_dir: PathBuf,
+    pub tmp_dir: Utf8PathBuf,
+    pub data_dir: DataDir,
+    pub cache_dir: Utf8PathBuf,
+    pub file_transfer_dir: Utf8PathBuf,
+    pub firmware_dir: Utf8PathBuf,
```
Comment (didier-wenzek) [inline crates/extensions/c8y_firmware_manager/src/config.rs:27]: This should be only a `DataDir` as all these paths are derived from it.

Comment (rina23q) [reply crates/extensions/c8y_firmware_manager/src/config.rs]: Done

**Review by albinsuresh — APPROVED** [review]

**Pull request merged**